### DISCUSSION TITLE
A few questions

### DISCUSSION THREAD

**Discussion by darleybarreto** [neuronika/neuronika#60] [Ideas]
Hi folks,

First of all I would like to say what impressive work so far! I am a person with engineering and research background (ML/DL). I have been using PyTorch and TensorFlow for a while now, and I must say that the architecture is similar and feels like a smooth transition from PyTorch to neuronika, which is amazing. 

I am a bit familiar with the Julia ecosystem too, specially Flux, Zygote and JuliaGPU. This ecosystem has a few points that might be interesting to discuss:

1. Abstraction over the autodiff machinery
2. Abstraction over devices
3. (Future abstraction over operations/devices using KernelAbstractions)

What you folks think about having a small abstraction over the choice of the autodiff (separating from the framework itself)? How would Rust play along with GPUs (Nvidia, AMD)? For instance, JuliaGPU has a compiler and a GC to deal with allocation of memory in the device. Would it be easy to compile Rust-y kernels directly to the GPU?

Best wishes,

**Comment by frjnn** [comment]
Thanks.

I'm not familiar with Julia at all, but I'm always willing to expand my horizons, would you be so kind to provide me some articles to begin with?

Concerning gp-GPUS, frankly, I don't really know, neuronika is built on top of ndarray and we'll probably wait and see what happens.

**Comment by darleybarreto** [comment]
> I'm not familiar with Julia at all, but I'm always willing to expand my horizons, would you be so kind to provide me some articles to begin with?

#### GPUs
I don't know how "old" [this](https://www.youtube.com/watch?v=7Yq1UyncDNc) video is, but I would recommend giving a look. Also the official [tutorial](https://cuda.juliagpu.org/stable/tutorials/introduction/) definitely helps. 

#### Flux
[Deep Learning with Flux - A 60 Minute Blitz](https://fluxml.ai/tutorials/2020/09/15/deep-learning-flux.html). Afaik, Zygote is a bit less explored in tutorials, but there is a [paper](https://arxiv.org/abs/1810.07951). Either way, it's a possible autodiff for Flux, there's also an "old" one called [ForwardDiff](https://github.com/JuliaDiff/ForwardDiff.jl) and a "newer", which is not ready yet.

**Comment by kirawi** [comment]
GPU depends upon https://github.com/rust-ndarray/ndarray/issues/840

**Comment by frjnn** [comment]
> GPU depends upon [rust-ndarray/ndarray#840](https://github.com/rust-ndarray/ndarray/issues/840)

It also depends on the current state of the Rust ecosystem. I'd be eager to start working on GPU support.

**Reply by kirawi** [reply]
Update: https://www.reddit.com/r/rust/comments/qzv428/announcing_the_rust_cuda_project_an_ecosystem_of/

**Reply by frjnn** [reply]
Much interesting. I was actually waiting for it. Thanks.

**Reply by kirawi** [reply]
ArrayFire was mentioned as well: https://github.com/arrayfire/arrayfire-rust. It has the drawback of being harder to package because it's not fully Rust though.

**Reply by frjnn** [reply]
I'll probably buy a Jetson card and start prototyping soon. Rust CUDA seems promising to me as it is the more idiomatic approach currently available. We want to keep things in Rust as much as possible.

**Comment by kirawi** [comment]
Perhaps eventually https://www.reddit.com/r/rust/comments/qi7f32/media_100_rust_path_tracer_running_on_cpu_gpu/

**Comment by frjnn** [comment]
Update on GPU support: I'm currently working on a type safe wrapper for the latest version of cuDNN @Rust-GPU, at the same time a rust wrapper for cuBLAS is being worked on also @Rust-GPU. 

Things are not there yet but it's now a work in progress.
